pkg/service/service.go

```diff
@@ -1466,43 +1473,38 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
 						backend.L3n4Addr, err)
 				}
 				backends[i].ID = id
-				// Default backend state is active.
-				backends[i].State = lb.BackendStateActive
+				backends[i].Weight = backend.Weight
 				newBackends = append(newBackends, backends[i])
 				// TODO make backendByHash by value not by ref
 				s.backendByHash[hash] = &backends[i]
 			} else {
 				backends[i].ID = s.backendByHash[hash].ID
-				backends[i].State = s.backendByHash[hash].State
 			}
 			svc.backendByHash[hash] = &backends[i]
 		} else {
 			backends[i].ID = b.ID
-			// Update backend state.
-			if b.RestoredFromDatapath {
-				backends[i].State = b.State
-				// Toggle the flag as the backend is now restored.
-				b.RestoredFromDatapath = false
-			} else {
-				// Backend state can either be updated via kubernetes events,
-				// or service API. If the state update is coming via kubernetes events,
-				// then we need to update the internal state. Currently, the only state
-				// update in this case is for the terminating state. All other state
-				// updates happen via the API (UpdateBackendState) in which case we need
-				// to set the backend state to the saved state.
-				if backends[i].State == lb.BackendStateTerminating &&
-					b.State != lb.BackendStateTerminating {
-					b.State = backends[i].State
-					// Update the persisted backend state in BPF maps.
-					if err := s.lbmap.UpdateBackendWithState(backends[i]); err != nil {
-						return nil, nil, nil, fmt.Errorf("failed to update backend %+v %w",
-							backends[i], err)
-					}
-				} else {
-					// Set the backend state to the saved state.
-					backends[i].State = b.State
+			b.RestoredFromDatapath = false
+			// Backend state can either be updated via kubernetes events,
+			// or service API. If the state update is coming via kubernetes events,
+			// then we need to update the internal state. Currently, the only state
+			// update in this case is for the terminating state or when backend
+			// weight has changed or is set to zero (maintenance). All other state
+			// updates happen via the API (UpdateBackendsState) in which case we need
+			// to set the backend state to the saved state.
+			if (backends[i].State == lb.BackendStateTerminating &&
+				b.State != lb.BackendStateTerminating) ||
+				b.Weight == 0 || b.Weight != backends[i].Weight {
```

Comment:
What the weight related checks are added to the code that updates persisted state? Is this a leftover from previous iterations where you were persisting backend weights in BPF maps?

Reply:
this is similar to the comment below.. 
Let's say I have existing backend. The agent has been restarted. I now want to set a weight of a backend to 0 for a specific service, but as I am not modifying backend state across all services the state would not be changed if the `||
b.Weight == 0` was not there and the latter wouldn't work as the backend weight is not recovered from anywhere.

Reply:
Is there a better way to handle these updates without overloading this code path with weights related checks? If we are not persisting backend weights, why do we want to persist the backend states derived from weight related API updates? We would want to have a consistent behavior across k8s and standalone LB modes. 
What do you think of updating the weights once all the services and their backends have been restored from the datapath - https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L869?

Reply:
I mean I could just remove this part if you think it's no good. We will use `cilium service update --backends "10.246.3.34:7047" --states active` anyway (after or before agent restart) to put backend out for maintenance.

Reply:
> What do you think of updating the weights once all the services and their backends have been restored from the datapath - https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L869?

I don't believe this makes sense. At this point, in func `RestoreServices()` we have zero knowledge about backend weights. So we are really just waiting for user to update a specific service and/or their backend weights.

Reply:
>So we are really just waiting for user to update a specific service and/or their backend weights.

What do you mean by this? Users are expected to set backend weights after agent restart. I'm suggesting that we block `cilium service` API until services are restored. In case of K8s, the agent removes stale services (SyncWithK8sFinished) only after all the services are restored, and sync with k8s api server is complete. Likewise for the standalone LB mode, you need to make sure that new cilium service API calls are done only after the restore operation is complete. 

> We will use `cilium service update --backends "10.246.3.34:7047" --states active` anyway (after or before agent restart) to put backend out/back for/from maintenance.

If you are going to anyway use the `--states` API, then in what cases is this code path needed to persist backend states derived from weights related updates?

Reply:
> We will use `cilium service update --backends "10.246.3.34:7047" --states active` anyway (after or before agent restart) to put backend out/back for/from maintenance.

If you are going to anyway use the `--states` API, then in what cases is this code path needed to persist backend states derived from weights related updates?

> So we are really just waiting for user to update a specific service and/or their backend weights.

What do you mean by this? Aren't users expected to set backend weights after agent restart anyway because the weights are not persisted?

Reply:
> If you are going to anyway use the --states API, then in what cases is this code path needed to persist backend states derived from weights related updates?

Well only for the case, when backend is put e.g. into maintenance, cilium-agent restarted, and user would want to put this backend into active state **only** for specific service (not for all). 

> What do you mean by this? Aren't users expected to set backend weights after agent restart anyway because the weights are not persisted?

No they are not expected to set backend weights after agent restart as they do not have to. If they haven't changed the maglev lookup table size the lut stays intact even after agent restart and the backend weights are only needed to modify the maglev lut.

Reply:
>No they are not expected to set backend weights after agent restart as they do not have to.

If you don't update weights, the second time the agent restarts, the agent will set backend weights to 0, and recalculate maglev table accordingly?

Reply:
We need to have a common logic that works for both k8s and standalone modes. Here are the code changes I propose  - 

- Add a flag to track whether backend weights are explicitly set. Set this flag whenever weights are set/updated for backends in the API handler. 
- Read this flag to determine the case where weights are updated, hence the saved backend weight and state need to be updated. 
- We can get rid of the `RestoredFromDatapath` flag, but we don't have to do that in your PR.

Here is the diff (doesn't include the API handler changes, but should be trivial):

<details><summary>proposed diff</summary>

diff --git a/pkg/loadbalancer/loadbalancer.go b/pkg/loadbalancer/loadbalancer.go
index dfdc77b1ae..c4275b7d01 100644
--- a/pkg/loadbalancer/loadbalancer.go
+++ b/pkg/loadbalancer/loadbalancer.go
@@ -310,7 +310,8 @@ type Backend struct {
        // ID of the backend
        ID BackendID
        // Weight of backend
-       Weight uint16
+       Weight        uint16
+       WeightUpdated bool
        // Node hosting this backend. This is used to determine backends local to
        // a node.
        NodeName string
diff --git a/pkg/service/service.go b/pkg/service/service.go
index dac44636cd..cb3848138d 100644
--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -1483,7 +1483,7 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
                        svc.backendByHash[hash] = &backends[i]
                } else {
                        backends[i].ID = b.ID
-                       b.RestoredFromDatapath = false
+                       updateSavedBeState := false
                        // Backend state can either be updated via kubernetes events,
                        // or service API. If the state update is coming via kubernetes events,
                        // then we need to update the internal state. Currently, the only state
@@ -1491,9 +1491,15 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
                        // weight has changed or is set to zero (maintenance). All other state
                        // updates happen via the API (UpdateBackendsState) in which case we need
                        // to set the backend state to the saved state.
-                       if (backends[i].State == lb.BackendStateTerminating &&
-                               b.State != lb.BackendStateTerminating) ||
-                               b.Weight == 0 || b.Weight != backends[i].Weight {
+                       switch {
+                       case backends[i].State == lb.BackendStateTerminating &&
+                               b.State != lb.BackendStateTerminating:
+                               updateSavedBeState = true
+                       case backends[i].WeightUpdated:
+                               updateSavedBeState = true
+                               b.Weight = backends[i].Weight
+                       }
+                       if updateSavedBeState {
                                b.State = backends[i].State
                                // Update the persisted backend state in BPF maps.
                                if err := s.lbmap.UpdateBackendWithState(backends[i]); err != nil {
@@ -1504,7 +1510,6 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
                                // Set the backend state to the saved state.
                                backends[i].State = b.State
                        }
-                       b.Weight = backends[i].Weight
                }
        }

</details>